File: scripts/dev_server.py
import sys
import os

try:
    import http.server
    import socketserver
except ModuleNotFoundError as e:
    print(f"Error: Required Python standard library module not found: {e}")
    print("Ensure you have Python 3.6 or later installed with the full standard library.")
    print("On Debian/Ubuntu: sudo apt install python3")
    sys.exit(1)

import errno

# Default PORT is 8080, can be overridden by first argument
PORT = int(sys.argv[1]) if len(sys.argv) > 1 else 8080

# Mapping of Clean URL Slug -> Real HTML File
route_map = {
    "/": "Home.html",
    "/about": "About.html",
    "/legal-framework": "Legal_Framework.html",
    "/updates": "The_Chronicle.html",
    "/connect": "Connect.html",
    "/support": "Support.html",
    "/privacy": "Privacy.html",
    "/reports": "Reports.html",
    "/grievance": "Grievance.html"
}

# Mapping of Real File -> Clean Slug (for 301 upgrades) home special case handled separately
file_to_slug = {v: k for k, v in route_map.items()}

class NSCWCleanURLServer(http.server.SimpleHTTPRequestHandler):
    def log_message(self, format, *args):
        # Console logging
        sys.stderr.write("%s - [%s] %s\n" % (self.address_string(), self.log_date_time_string(), format%args))

    def do_GET(self):
        # 1. Handle Clean URL Slug -> HTML file mapping (Internal Rewrite)
        parts = self.path.split('?')
        clean_path = parts[0].rstrip('/')
        if clean_path == "":
            clean_path = "/"
        
        # 2. Redirect .html requests to clean URLs (301)
        if clean_path.endswith('.html'):
            filename = os.path.basename(clean_path)
            if filename in file_to_slug:
                new_slug = file_to_slug[filename]
                if len(parts) > 1:
                    new_slug += "?" + parts[1]
                
                print(f"301 Redirecting: {self.path} -> {new_slug}")
                self.send_response(301)
                self.send_header('Location', new_slug)
                self.end_headers()
                return

        # 3. Handle Clean URL Slug -> HTML file mapping (Internal Rewrite)
        if clean_path in route_map:
            target_file = route_map[clean_path]
            print(f"Serving Clean URL: {self.path} via {target_file}")
            
            try:
                with open(target_file, 'rb') as f:
                    content = f.read()
                self.send_response(200)
                self.send_header("Content-type", "text/html")
                self.end_headers()
                self.wfile.write(content)
                return
            except FileNotFoundError:
                # If the mapped file is missing, fall through to 404
                pass

        # 4. Check if it's a physical file (CSS, JS, Images, etc.)
        if os.path.exists(self.translate_path(self.path)) and not os.path.isdir(self.translate_path(self.path)):
            super().do_GET()
            return

        # 5. Fallback: Serve custom 404.html
        print(f"404 Not Found: {self.path} -> Serving 404.html")
        try:
            with open("404.html", 'rb') as f:
                content = f.read()
            self.send_response(404)
            self.send_header("Content-type", "text/html")
            self.end_headers()
            self.wfile.write(content)
        except FileNotFoundError:
            self.send_error(404, "404 Page Not Found")

    def do_HEAD(self):
        super().do_HEAD()

if __name__ == "__main__":
    # Change working directory to public/
    if os.path.exists("public"):
        os.chdir("public")
    
    # Create the server
    socketserver.TCPServer.allow_reuse_address = True
    try:
        with socketserver.TCPServer(("", PORT), NSCWCleanURLServer) as httpd:
            print("\n" + "="*50)
            print(f"NSCW Web Platform Dev Server (Clean URLs)")
            print(f"Local URL: http://localhost:{PORT}")
            print("="*50)
            print("Configured Routes:")
            for slug, file in route_map.items():
                print(f"  {slug.ljust(18)} -> {file}")
            print("="*50 + "\n")
            sys.stdout.flush()
            httpd.serve_forever()
    except OSError as e:
        if e.errno == errno.EADDRINUSE:
            print(f"\n[ERROR] Port {PORT} is already in use.")
            print(f"Try running with a different port: python3 dev_server.py 8085")
            print(f"Or kill the existing process: lsof -ti:{PORT} | xargs kill -9")
        else:
            raise e
    except KeyboardInterrupt:
        print("\nShutting down server.")

File: scripts/dev_server.js
#!/usr/bin/env node
'use strict';

const http = require('http');
const fs = require('fs');
const path = require('path');

const PORT = parseInt(process.argv[2] || '8080', 10);

const routeMap = {
  '/': 'Home.html',
  '/about': 'About.html',
  '/legal-framework': 'Legal_Framework.html',
  '/updates': 'The_Chronicle.html',
  '/connect': 'Connect.html',
  '/support': 'Support.html',
  '/privacy': 'Privacy.html',
  '/reports': 'Reports.html',
  '/grievance': 'Grievance.html',
  '/mandate': 'Mandate.html',
};

const fileToSlug = Object.fromEntries(
  Object.entries(routeMap).map(([slug, file]) => [file, slug])
);

const mimeTypes = {
  '.html': 'text/html',
  '.css': 'text/css',
  '.js': 'application/javascript',
  '.json': 'application/json',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.gif': 'image/gif',
  '.svg': 'image/svg+xml',
  '.ico': 'image/x-icon',
  '.woff': 'font/woff',
  '.woff2': 'font/woff2',
  '.ttf': 'font/ttf',
  '.otf': 'font/otf',
  '.webp': 'image/webp',
  '.pdf': 'application/pdf',
};

const server = http.createServer((req, res) => {
  const [rawPath, query] = req.url.split('?');
  const qs = query ? '?' + query : '';
  const cleanPath = rawPath.replace(/\/+$/, '') || '/';

  const now = new Date().toUTCString();
  const log = (msg) => process.stderr.write(`${req.socket.remoteAddress} - [${now}] ${msg}\n`);

  // 301 redirect .html requests to clean URLs
  if (cleanPath.endsWith('.html')) {
    const filename = path.basename(cleanPath);
    if (fileToSlug[filename]) {
      const newSlug = fileToSlug[filename] + qs;
      log(`301 Redirecting: ${req.url} -> ${newSlug}`);
      res.writeHead(301, { Location: newSlug });
      res.end();
      return;
    }
  }

  // Serve clean URL slug -> HTML file
  if (routeMap[cleanPath]) {
    const targetFile = routeMap[cleanPath];
    process.stdout.write(`Serving Clean URL: ${req.url} via ${targetFile}\n`);
    try {
      const content = fs.readFileSync(targetFile);
      res.writeHead(200, { 'Content-Type': 'text/html' });
      res.end(content);
      return;
    } catch {
      // fall through to 404
    }
  }

  // Serve physical files (CSS, JS, images, etc.)
  // Use realpathSync to resolve symlinks and guard against path traversal.
  const root = path.resolve(process.cwd());
  const candidates = [
    path.resolve(root, '.' + cleanPath),
    path.resolve(root, 'public', '.' + cleanPath),
  ];
  try {
    const candidate = candidates.find((p) => { try { return fs.statSync(p).isFile(); } catch { return false; } });
    if (!candidate) throw new Error('not found');
    const realFilePath = fs.realpathSync(candidate);
    const relative = path.relative(root, realFilePath);
    if (relative && !relative.startsWith('..') && !path.isAbsolute(relative) && fs.statSync(realFilePath).isFile()) {
      const ext = path.extname(realFilePath).toLowerCase();
      const contentType = mimeTypes[ext] || 'application/octet-stream';
      const content = fs.readFileSync(realFilePath);
      res.writeHead(200, { 'Content-Type': contentType });
      res.end(content);
      return;
    }
  } catch {
    // path doesn't exist or is outside root — fall through to 404
  }

  // Fallback: serve 404.html
  process.stdout.write(`404 Not Found: ${req.url} -> Serving 404.html\n`);
  try {
    const content = fs.readFileSync('404.html');
    res.writeHead(404, { 'Content-Type': 'text/html' });
    res.end(content);
  } catch {
    res.writeHead(404, { 'Content-Type': 'text/plain' });
    res.end('404 Page Not Found');
  }
});

server.on('error', (err) => {
  if (err.code === 'EADDRINUSE') {
    console.error(`\n[ERROR] Port ${PORT} is already in use.`);
    console.error(`Try running with a different port: node dev_server.js 8085`);
    console.error(`Or kill the existing process using your OS process manager.`);
  } else {
    throw err;
  }
  process.exit(1);
});

// Change working directory to public/ if it exists
if (fs.existsSync('public')) {
  process.chdir('public');
}

server.listen(PORT, () => {
  const separator = '='.repeat(50);
  console.log('\n' + separator);
  console.log('NSCW Web Platform Dev Server (Clean URLs)');
  console.log(`Local URL: http://localhost:${PORT}`);
  console.log(separator);
  console.log('Configured Routes:');
  for (const [slug, file] of Object.entries(routeMap)) {
    console.log(`  ${slug.padEnd(18)} -> ${file}`);
  }
  console.log(separator + '\n');
});

process.on('SIGINT', () => {
  console.log('\nShutting down server.');
  server.close(() => process.exit(0));
});
